sensor_msgs_py/sensor_msgs_py/point_cloud2.py

```diff
@@ -157,28 +228,52 @@ def create_cloud(header, fields, points):
                    the fields parameter)
     :return: The point cloud as sensor_msgs.msg.PointCloud2
     """
-    cloud_struct = struct.Struct(_get_struct_fmt(False, fields))
-
-    buff = ctypes.create_string_buffer(cloud_struct.size * len(points))
-
-    point_step, pack_into = cloud_struct.size, cloud_struct.pack_into
-    offset = 0
-    for p in points:
-        pack_into(buff, offset, *p)
-        offset += point_step
-
-    return PointCloud2(header=header,
-                       height=1,
-                       width=len(points),
-                       is_dense=False,
-                       is_bigendian=False,
-                       fields=fields,
-                       point_step=cloud_struct.size,
-                       row_step=cloud_struct.size * len(points),
-                       data=buff.raw)
-
-
-def create_cloud_xyz32(header, points):
+    # Check if input is numpy array
+    if isinstance(points, np.ndarray):
+        # Check if this is an unstructured array
+        if points.dtype.names is None:
+            assert all(fields[0].datatype == field.datatype for field in fields[1:]), \
+                "All fields need to have the same datatype. Pass a structured NumPy array \
+                    with multiple dtypes otherwise."
+            # Convert unstructured to structured array
+            points = unstructured_to_structured(
+                points,
+                dtype=dtype_from_fields(fields))
+        else:
+            assert points.dtype == dtype_from_fields(fields), \
+                "PointFields and structured NumPy array dtype do not match for all fields! \
+                    Check their field order, names and types."
+    else:
+        # Cast python objects to structured NumPy array (slow)
+        points = np.array(
+            # Points need to be tuples in the structured array
+            list(map(tuple, points)),
+            dtype=dtype_from_fields(fields))
+
+    # Handle organized clouds
+    assert len(points.shape) <= 2, \
+        "Too many dimensions for organized cloud! \
+            Points can only be organized in max. two dimensional space"
+    height = 1
+    width = points.shape[0]
+    # Check if input points are an organized cloud (2D array of points)
+    if len(points.shape) == 2:
+        height = points.shape[1]
+
+    # Put everything together
+    return PointCloud2(
+        header=header,
+        height=height,
+        width=width,
+        is_dense=False,
+        is_bigendian=sys.byteorder != 'little',
+        fields=fields,
+        point_step=points.dtype.itemsize,
+        row_step=(points.dtype.itemsize * width),
+        data=points.tobytes())
```

Comment:
@Flova I just spent a while improving performance for PointCloud2 message creation on the ros2_numpy repo, and thought I'd pass on some knowledge. I noticed that  doing `data=np_array.tobytes()` can be quite slow on large point clouds. I have some code in [this PR](https://github.com/Box-Robotics/ros2_numpy/pull/3) using a memoryview and `array.array.frombytes()` which I found to be 115x faster.

Let me know what you think- there's some overlap in what we're working on I think.

Here's an example of how to do the same in this repo:
```suggestion
    memory_view = memoryview(points)
    if memory_view.nbytes > 0:
        array_bytes = memory_view.cast("B")
    else:
        # Casting raises a TypeError if the array has no elements
        array_bytes = b""
    as_array = array.array("B")
    as_array.frombytes(array_bytes)

    # Put everything together
    return PointCloud2(
        header=header,
        height=height,
        width=width,
        is_dense=False,
        is_bigendian=sys.byteorder != 'little',
        fields=fields,
        point_step=points.dtype.itemsize,
        row_step=(points.dtype.itemsize * width),
        data=as_array)
```

Reply:
I really liked ros_numpy in ROS1. Cool that it is still a thing in ROS2.

I did some performance testing as seen above. I did not noticed speeds in the seconds as described in your PR. Even for very large point clouds (~7GB), but this could also be related to the used hardware. This being said I also wondered about a multiple magnitude difference in serialisation vs. deserialization. With the serialisation being significantly slower (ms instead of us).

We also kind of do the same thing. You are right with this. But I don't know if the maintainers want to add ros_numpy as a dependency.

Reply:
> We also kind of do the same thing. You are right with this. But I don't know if the maintainers want to add ros_numpy as a dependency.

I'd rather not, both to keep the number of dependencies down, and also to avoid circular dependency problems (it looks like ros2_numpy depends on nav_msgs, which is in this repository).  Can we just use portions of the code that are relevant?

Reply:
Okay. I think both ones are very similar now, but I had the idea to make a PR to ros2_numpy to port so minor stuff from here to there (Mainly dropping the `pftype_sizes` mapping as it can be inferred from the dtype).

Reply:
I tested the proposed changes on my machine (ROS2 rolling, Ubuntu 20.04, Ryzen 1700X) and could not reproduce the described behavior. It was even slightly slower:

This PR:
```python
points = np.random.rand(1080*1920, 3).astype(np.float32)

In [84]: points = unstructured_to_structured(points, names=["x", "y", "z"])

In [85]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 9.11 ms, sys: 7.67 ms, total: 16.8 ms
Wall time: 17.2 ms

In [86]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 1.76 ms, sys: 8.04 ms, total: 9.8 ms
Wall time: 9.78 ms

In [87]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 4.15 ms, sys: 14 µs, total: 4.16 ms
Wall time: 4.18 ms

In [88]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 4.17 ms, sys: 18 µs, total: 4.19 ms
Wall time: 4.21 ms

In [89]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 4.27 ms, sys: 32 µs, total: 4.31 ms
Wall time: 4.33 ms

In [90]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 4.18 ms, sys: 19 µs, total: 4.2 ms
Wall time: 4.23 ms
```

Proposed changes:
```python
In [5]: points = np.random.rand(1080*1920, 3).astype(np.float32)

In [6]: points = unstructured_to_structured(points, names=["x", "y", "z"])

In [7]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 12.8 ms, sys: 8.48 ms, total: 21.2 ms
Wall time: 20.5 ms

In [8]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 11.1 ms, sys: 7.56 ms, total: 18.7 ms
Wall time: 18.6 ms

In [9]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 156 µs, sys: 12.4 ms, total: 12.6 ms
Wall time: 12.5 ms

In [10]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 3.86 ms, sys: 12.2 ms, total: 16 ms
Wall time: 15.9 ms

In [11]: %time a = create_cloud_xyz32(Header(), points)
CPU times: user 6.83 ms, sys: 3.31 ms, total: 10.1 ms
Wall time: 10.1 ms
```

Reply:
This also happens with different arrays and dtype combinations.

Reply:
```python
    def __init__(self, **kwargs):
        assert all('_' + key in self.__slots__ for key in kwargs.keys()), \
            'Invalid arguments passed to constructor: %s' % \
            ', '.join(sorted(k for k in kwargs.keys() if '_' + k not in self.__slots__))
        from std_msgs.msg import Header
        self.header = kwargs.get('header', Header())
        self.height = kwargs.get('height', int())
        self.width = kwargs.get('width', int())
        self.fields = kwargs.get('fields', [])
        self.is_bigendian = kwargs.get('is_bigendian', bool())
        self.point_step = kwargs.get('point_step', int())
        self.row_step = kwargs.get('row_step', int())
        self.data = array.array('B', kwargs.get('data', []))
        self.is_dense = kwargs.get('is_dense', bool())
```

This is the PointCloud constructor on my installation. Like you described, but without the performance implications.

Reply:
The problem is that the `self.data = ...` line is actually calling a property for the data setter:

```python
class PointCloud2:
	...

	@data.setter
    def data(self, value):
        if isinstance(value, array.array):
            assert value.typecode == 'B', \
                "The 'data' array.array() must have the type code of 'B'"
            self._data = value
            return
        self._data = array.array('B', value)
```

^ This is what it looks like for foxy. The `array.array('B, value)` line is quite expensive, at least from my testing. Doing `array.array.frombytes(a_memoryview)` was faster. I'm wondering if `rolling` improved this setter logic to be more efficient for bytes datatypes.

Reply:
I agree with you, but my code snippet is sufficient, if you only use the constructor. It also includes the  `array.array('B', value)` part, so it is skipped later on by the setter in the best case. But the interesting part `array.array('B', value)` is included.

Reply:
Ahh my apologies, I think I got confused. You're right- your code snippet is running `array.array('B, value)`. The difference is that the implementation for `array.array.frombytes(memoryview)` does a simple `memcpy` for the entire buffer. ([implementation of frombytes](https://github.com/python/cpython/blob/434ffb7f1f86e6b0cdfad3ede59993934d86e464/Modules/arraymodule.c#L1682))

This is different than the `array.array('B', value)` [implementation](https://github.com/python/cpython/blob/434ffb7f1f86e6b0cdfad3ede59993934d86e464/Modules/arraymodule.c#L2630) which is more complicated, and seems to be slower (from my testing). 

Either way, the memoryview implementation is only one copy: the `memcpy`, whereas doing `np.tobytes()` is one copy, and then `array.array('B", value)` performs another copy, and iterates over the data which is added cost.

Reply:
I agree that there are happening two copies, but for now I could not measure any performance benefit. Could you try to reproduce your tests describing the exact configuration of your data and system?

Reply:
Okay, I had a chance to do some quick tests: https://drive.google.com/drive/u/0/folders/1kxmAn1ND_HF_eCCpH2kXz1V78ajQnP6D

I created a minimal reproducible benchmark that doesn't include ros2_numpy or the PointCloud2 messages. I was curiously unable to reproduce the extremely slow serialization that I saw with ros2_numpy- instead, I saw only a 51% increase.

My theory is that something about how ros2_numpy creates the point cloud ([pointcloud2_to_array](https://github.com/Box-Robotics/ros2_numpy/blob/238b6a2d66d04fb687ce421cc6a7ce63428bbaa3/ros2_numpy/point_cloud2.py#L134)) might be creating an array that's somehow slow to call `tobytes()` on... I think I need to do some more investigating on my end.

 These are the results on my machine:
```
Implementation 1: 2.3597640991210938 seconds
Implementation 2: 1.2081842422485352 seconds
```

Reply:
I tested your minimal example, and it was indeed faster. But now comes the funny part. It was exactly equal if I added the `PointCloud2` message again!
I dug deeper and the issue lies in the implementation of the constructor. Which has a redundant typecast. If you use the property directly, it is similar to your implementation.

This therefore seems to be an upstream issue. We should open an issue there and use your proposed changes.

Timing:
```bash
Implementation 1: 2.7484822273254395 seconds   # This PR
Implementation 2: 1.379800796508789 seconds    # Property with your cast
Implementation 3: 2.714930295944214 seconds    # Constructor (as in this PR) with your cast
```

<details>
  <summary>Code</summary>

  ```python
import array
from time import time
from contextlib import contextmanager
from sensor_msgs.msg import PointCloud2, PointField
import numpy as np
import sys
from std_msgs.msg import Header


@contextmanager
def timer(name: str):
    start_time = time()
    yield
    print(f"{name}: {time() - start_time} seconds")


def perform_tests():
    samples = 100
    np_cloud = np.load("np_cloud.npy")
    # dtype: [('x', '<f4'), ('y', '<f4'), ('z', '<f4'), ('a', 'u1'), ('rgb', '<f4')]

    with timer("Implementation 1"):
        for _ in range(samples):
            fields = [PointField(name='x', offset=0,
                         datatype=PointField.FLOAT32, count=1),
              PointField(name='y', offset=4,
                         datatype=PointField.FLOAT32, count=1),
              PointField(name='z', offset=8,
                         datatype=PointField.FLOAT32, count=1)]
            c = PointCloud2(
                header=Header(),
                height=1,
                width=len(np_cloud),
                is_dense=False,
                is_bigendian=sys.byteorder != 'little',
                fields=fields,
                point_step=np_cloud.dtype.itemsize,
                row_step=(np_cloud.dtype.itemsize * len(np_cloud)),
                data=np_cloud.tobytes())

    with timer("Implementation 2"):
        for _ in range(samples):
            fields = [PointField(name='x', offset=0,
                         datatype=PointField.FLOAT32, count=1),
              PointField(name='y', offset=4,
                         datatype=PointField.FLOAT32, count=1),
              PointField(name='z', offset=8,
                         datatype=PointField.FLOAT32, count=1)]
            
            memory_view = memoryview(np_cloud)
            casted = memory_view.cast("B")
            array_array = array.array("B")
            array_array.frombytes(casted)
            
            c = PointCloud2(
                header=Header(),
                height=1,
                width=len(np_cloud),
                is_dense=False,
                is_bigendian=sys.byteorder != 'little',
                fields=fields,
                point_step=np_cloud.dtype.itemsize,
                row_step=(np_cloud.dtype.itemsize * len(np_cloud)))
            c.data = array_array


    with timer("Implementation 3"):
        for _ in range(samples):
            fields = [PointField(name='x', offset=0,
                         datatype=PointField.FLOAT32, count=1),
              PointField(name='y', offset=4,
                         datatype=PointField.FLOAT32, count=1),
              PointField(name='z', offset=8,
                         datatype=PointField.FLOAT32, count=1)]
            
            memory_view = memoryview(np_cloud)
            casted = memory_view.cast("B")
            array_array = array.array("B")
            array_array.frombytes(casted)
            
            c = PointCloud2(
                header=Header(),
                height=1,
                width=len(np_cloud),
                is_dense=False,
                is_bigendian=sys.byteorder != 'little',
                fields=fields,
                point_step=np_cloud.dtype.itemsize,
                row_step=(np_cloud.dtype.itemsize * len(np_cloud)),
                data = array_array)


perform_tests()

```
  
</details>

Reply:
I opened #177 and #176 for these issues.

Reply:
This is amazing, thank you for the in-depth debugging. I had no idea that `__debug__` was on by default on most machines. That's definitely the cause of the extremely slow serialization I was seeing in https://github.com/Box-Robotics/ros2_numpy/pull/3!